spec/runtime/loader.spec.ts

```diff
@@ -310,6 +310,11 @@ describe('loadStack', () => {
           },
         },
       },
+      {
+        name: 'has params',
+        modulePath: './spec/fixtures/sources/commonjs-params',
+        expected: { ...expected, params: [{ name: 'FOO', type: 'string' }] },
```

Comment:
@mbleigh It still feels wrong to me that a param has a discriminating union field and input doesn't.

Why is it

```
{
  name: "FOO",
  type: "string",
  input: {
    textInput: {}
  }
}
```

and not

```
{
  name: "FOO",
  type: "string",
  input: {
    type: "text",
  }
}
```

or even
```
{
  stringParam: {
    name: "FOO",
    type: "string",
    input: {
      textInput: {}
    }
  }
}
```

It seems like those would be more ideologically consistent.

Reply:
Because one is defining a small enumerated set of types to which a string value can be cast, and the other is defining an unbounded set of configuration and validation hints for client UIs.

The main reason I care is because I don't want us to end up with a giant grab-bag of fields with insanely overcomplicated "if type is X then fields A, B, D, and F are applicable. if type is Y then fields A, D, F, G, and K are applicable".

The cast type has no additional configuration, so it works fine as a simple enum. Likewise if the input type had no additional configuration, it would be fine as a simple enum. But we aim to have many different types of inputs (checkbox, radio group, dropdown, input, text field, resource picker, secret picker, more as we need) and each of those will have some amount of configuration.

I look at a `oneof` as basically being an "enum with configuration", which is why I don't find this API inconsistent.

Reply:
I like Scala's term for "enum with configuration"--case class.

I see some of your argument, though the last example (where each is a oneof message) follows the argument even further. In the proposal, all input types can be used with all parameter types, but this isn't true. As I understand the current proposal, multiselect input can only be used with a list param (which can only use a multiselect input I think). And a text input can only use certain validators depending on their param type. While the last format makes my stomach churn instinctively, it has the benefit that the type system only allows the right input to be picked with the right param type.